Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,418 @@
+        await messagesSequenceDelegate.setConsumer(self)
+    }
+
+    /// Initialize a new ``KafkaConsumer`` and subscribe to the given list of `topics` as part of
+    /// the consumer group specified in `groupID`.
+    /// - Parameter topics: An array of topic names to subscribe to.
+    /// - Parameter groupID: Name of the consumer group that this ``KafkaConsumer`` will create / join.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the ``KafkaConsumer``.
+    /// - Parameter logger: A logger.
+    public convenience init(
+        topics: [String],
+        groupID: String,
+        config: KafkaConfig = KafkaConfig(),
+        logger: Logger
+    ) async throws {
+        // TODO: make prettier
+        // TODO: use inout parameter here?
+        var config = config
+        if let configGroupID = config.value(forKey: "group.id") {
+            if configGroupID != groupID {
+                throw KafkaError(description: "Group ID does not match with group ID found in the configuration")
+            }
+        } else {
+            try config.set(groupID, forKey: "group.id")
+        }
+
+        try await self.init(
+            config: config,
+            logger: logger
+        )
+        try await self.subscribe(topics: topics)
+    }
+
+    /// Initialize a new ``KafkaConsumer`` and assign it to a specific `partition` of a `topic`.
+    /// - Parameter topic: Name of the topic that this ``KafkaConsumer`` will read from.
+    /// - Parameter partition: Partition that this ``KafkaConsumer`` will read from.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the ``KafkaConsumer``.
+    /// - Parameter logger: A logger.
+    /// - Note: This consumer ignores the `group.id` property of its `config`.
+    public convenience init(
+        topic: String,
+        partition: KafkaPartition,
+        config: KafkaConfig = KafkaConfig(),
+        logger: Logger
+    ) async throws {
+        // Althogh an assignment is not related to a consumer group,
+        // librdkafka requires us to set a `group.id`.
+        // This is a known issue:
+        // https://github.com/edenhill/librdkafka/issues/3261
+        var config = config
+        try config.set(UUID().uuidString, forKey: "group.id")
+
+        try await self.init(
+            config: config,
+            logger: logger
+        )
+        try await self.assign(topic: topic, partition: partition)
+    }
+
+    deinit {
+        switch self.state {
+        case .started:
+            Task {
+                try await self._close()
+            }
+        case .closed:
+            return
+        }
+    }
+
+    /// Subscribe to the given list of `topics`.
+    /// The partition assignment happens automatically using `KafkaConsumer`'s consumer group.
+    /// - Parameter topics: An array of topic names to subscribe to.
+    private func subscribe(topics: [String]) throws {
+        // TODO: is this state needed for a method that is only invoked upon init?
```

Comment:
Currently, `subscribe` is only invoked by `init`. Thus we could potentially get rid of the state. But this totally depends on if we want to allow users to `subscribe/assign` to more topics during runtime.